README.md
# SACPY -- A Python Package for Statistical Analysis of Climate

**Sacpy**, an effecient Statistical Analysis tool for Climate or Meteorology data.

Author : Zilu Meng 

e-mail : zilumeng@uw.edu

github : https://github.com/ZiluM/sacpy

pypi : https://pypi.org/project/sacpy/

Document: https://zilum.github.io/sacpy/

version : 0.0.19

## Why choose Sacpy?

### Fast!

For example, Sacpy is more than 60 times faster than the traditional regression analysis with Python (see **speed test**). The following is the time spent performing the same task. Sacpy is fastest.

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/speed_test_00.png)

### Turn to climate data customization!

Compatible with commonly used meteorological calculation libraries such as numpy and xarray.

### Concise code

You can finish drawing a following figure with just seven lines of code. see **examples of concise**.

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/one_test.png)

You can use SVD/MCA to get the image below easily.

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/SVD.png)

## Install and update

You can use pip to install.

    pip install sacpy

Or you can visit https://gitee.com/zilum/sacpy/tree/main/dist to download **.whl file**, then

    pip install .whl_file

update:

    pip install --upgrade sacpy

or you can download **.whl** file and then install use ` pip install .whl_file`.

## Speed

As a comparison, we use the  **corr**  function in the xarray library, **corrcoef** function in numpy library, cdist in scipy, apply_func in xarray  and **for-loop**. The time required to calculate the correlation coefficient between SSTA and nino3.4 for 50 times is shown in the figure below.

It can be seen that we are four times faster than scipy cdist, five times faster than xarray.corr, 60 times faster than forloop, 110 times faster than xr.apply_func and 200 times faster than numpy.corrcoef.

Moreover, xarray and numpy can not return the **p value**. We can simply check the pvalue attribute of sacpy to get the p value.

All in all, if we want to get p-value and correlation or slope, we only to choose **Sacpy is 60 times faster than before**.

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/speed_test_00.png)

## Example

### example1

Calculate the correlation between SST and nino3.4 index

```Python
import numpy as np
import scapy as scp
import matplotlib.pyplot as plt
import sacpy.Map # need cartopy or you can just not import
import cartopy.crs as ccrs

# load sst
sst = scp.load_sst()['sst']
# get ssta (method=1, Remove linear trend;method=0, Minus multi-year average)
ssta = scp.get_anom(sst,method=1)
# calculate Nino3.4
Nino34 = ssta.loc[:,-5:5,190:240].mean(axis=(1,2))
# regression
linreg = scp.LinReg(Nino34,ssta)
# plot
fig = plt.figure(figsize=[7, 3])
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
lon ,lat = ssta.lon , ssta.lat
# shading
m = ax.scontourf(lon,lat,linreg.corr)
# significant plot
n = ax.sig_plot(lon,lat,linreg.p_value,color="k",marker="..")
# initialize map
ax.init_map(stepx=50, ysmall=2.5)
# colorbar
plt.colorbar(m)
# save
plt.savefig("../pic/nino34.png",dpi=200)

```

Result(For a detailed drawing process, see **example**):

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/nino34.png)

### example2

multiple linear regression on Nino3.4 IOD Index and ssta pattern

```Python
import numpy as np
import scapy as scp
import matplotlib.pyplot as plt


# load sst
sst = scp.load_sst()['sst']
# get ssta (method=1, Remove linear trend;method=0, Minus multi-year average)
ssta = scp.get_anom(sst,method=1)
# calculate Nino3.4
Nino34 = ssta.loc[:,-5:5,190:240].mean(axis=(1,2))
# calculate IODIdex
IODW = ssta.loc[:,-10:10,50:70].mean(axis=(1,2))
IODE = ssta.loc[:,-10:0,90:110].mean(axis=(1,2))
IODI = +IODW - IODE
# get x
X = np.vstack([np.array(Nino34),np.array(IODI)]).T
# multiple linear regression
MLR = scp.MultLinReg(X,ssta)

# plot IOD's effect
import sacpy.Map
import cartopy.crs as ccrs


fig = plt.figure(figsize=[7, 3])
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
lon ,lat = ssta.lon , ssta.lat
m = ax.scontourf(lon,lat,MLR.slope[1])
# significant plot
n = ax.sig_plot(lon,lat,MLR.pv_i[1],color="k",marker="..")
# initialize map
ax.init_map(stepx=50, ysmall=2.5)
plt.colorbar(m)
plt.savefig("../pic/MLR.png",dpi=200)
```

Result(For a detailed drawing process, see **example**):

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/MLR.png)

### example3

What effect will ENSO have on the sea surface temperature in the next summer?

```Python
import numpy as np
import sacpy as scp
import matplotlib.pyplot as plt
import xarray as xr

# load sst
sst = scp.load_sst()['sst']
ssta = scp.get_anom(sst)

# calculate Nino3.4
Nino34 = ssta.loc[:,-5:5,190:240].mean(axis=(1,2))

# get DJF mean Nino3.4
DJF_nino34 = scp.XrTools.spec_moth_yrmean(Nino34,[12,1,2])

# get JJA mean ssta
JJA_ssta = scp.XrTools.spec_moth_yrmean(ssta, [6,7,8])

# regression
reg = scp.LinReg(DJF_nino34[:-1], JJA_ssta)
# plot
import cartopy.crs as ccrs
import sacpy.Map


fig = plt.figure(figsize=[7, 3])
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
lon ,lat = np.array(ssta.lon) , np.array(ssta.lat)
m = ax.scontourf(lon,lat,reg.slope)
n = ax.sig_plot(lon,lat,reg.p_value,color="k",marker="///")
ax.init_map(stepx=50, ysmall=2.5)
plt.colorbar(m)
plt.savefig("../pic/ENSO_Next_year_JJA.png",dpi=300)

```

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/ENSO_Next_year_JJA.png)

Same as **Indian Ocean Capacitor Effect on Indo–Western Pacific Climate during the Summer following El Niño** (Xie et al.), the El Nino will lead to Indian ocean warming in next year JJA.

### example4

EOF analysis

```Python
import sacpy as scp
import numpy as np
import matplotlib.pyplot as plt
# get data
sst = scp.load_sst()["sst"].loc[:, -20:30, 150:275]
ssta = scp.get_anom(sst)
# EOF
eof = scp.EOF(np.array(ssta))
eof.solve()
# get spartial pattern and pc
pc = eof.get_pc(npt=2)
pt = eof.get_pt(npt=2)
# plot
import cartopy.crs as ccrs
import sacpy.Map
lon , lat = np.array(ssta.lon) , np.array(ssta.lat)
fig = plt.figure(figsize=[15,10])
ax = fig.add_subplot(221,projection=ccrs.PlateCarree(central_longitude=180))
m1 = ax.scontourf(lon,lat,pt[0,:,:],cmap='RdBu_r',levels=np.linspace(-0.75,0.75,15),extend="both")
ax.scontour(m1,colors="black")
ax.init_map(ysmall=2.5)
# plt.colorbar(m1)
ax2 = fig.add_subplot(222)
ax2.plot(sst.time,pc[0])
ax3 = fig.add_subplot(223,projection=ccrs.PlateCarree(central_longitude=180))
m2 = ax3.scontourf(lon,lat,pt[1,:,:],cmap='RdBu_r',levels=np.linspace(-0.75,0.75,15),extend="both")
ax3.scontour(m2,colors="black")
ax3.init_map(ysmall=2.5)
ax4 = fig.add_subplot(224)
ax4.plot(sst.time,pc[1])
cb_ax = fig.add_axes([0.1,0.06,0.4,0.02])
fig.colorbar(m1,cax=cb_ax,orientation="horizontal")
plt.savefig("../pic/eof_ana.png",dpi=300)
```

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/eof_ana.png)

## example5

Mean value (Composite Analysis) t-test for super El Nino (DJF Nino3.4 > 1)

```Python

import sacpy as scp
import numpy as np
import matplotlib.pyplot as plt

sst = scp.load_sst()["sst"]
ssta = scp.get_anom(sst, method=0)
# get Dec Jan Feb SSTA
ssta_djf = scp.XrTools.spec_moth_yrmean(ssta,[12,1,2])
Nino34 = ssta_djf.loc[:, -5:5, 190:240].mean(axis=(1, 2))
# select year of Super El Nino
select = Nino34 >= 1
ssta_sl = ssta_djf[select]
mean, pv = scp.one_mean_test(ssta_sl)
# plot
import sacpy.Map
import cartopy.crs as ccrs
fig = plt.figure(figsize=[7, 3])
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
lon ,lat = np.array(ssta.lon) , np.array(ssta.lat)
m = ax.scontourf(lon,lat,mean)
n = ax.sig_plot(lon,lat,pv,color="k",marker="..")
ax.init_map(stepx=50, ysmall=2.5)
plt.colorbar(m)
plt.savefig("../pic/one_test.png")
```

Result:

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/one_test.png)

## example6

SVD(MCA) analysis.

```Python
import sacpy as scp
import xarray as xr
import matplotlib.pyplot as plt
import numpy as np
from xmca import array
import sacpy.Map
import cartopy.crs as ccrs

# load data
sst = scp.load_sst()['sst'].loc["1991":"2021", -20:30, 150:275]
ssta = scp.get_anom(sst)
u = scp.load_10mwind()['u']
v = scp.load_10mwind()['v']

uua = scp.get_anom(u)
vua = scp.get_anom(v)
uv = np.concatenate([np.array(uua)[...,np.newaxis],np.array(vua)[...,np.newaxis]],axis=-1)
# calculation
svd = scp.SVD(ssta,uv,complex=False)
svd.solve()
ptl, ptr = svd.get_pt(3)
pcl,pcr = svd.get_pc(3)
upt ,vpt = ptr[...,0] , ptr[...,1]
sst_pt = ptl
# plot progress, see example/SVD.ipynb
```

result:

![](https://raw.githubusercontent.com/ZiluM/sacpy/main/pic/SVD.png)

## examples of concise

If you want to plot example1's figure , you need write:

```Python
from cartopy.mpl.ticker import LongitudeFormatter, LatitudeFormatter
from matplotlib.ticker import MultipleLocator
import cartopy.crs as ccrs
plt.rc('font', family='Times New Roman', size=12)
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
m = ax.contourf(ssta.lon,ssta.lat,linreg.corr,
                cmap="RdBu_r",
                levels=np.linspace(-1, 1, 15),
                extend="both",
                transform=ccrs.PlateCarree())
n = plt.contourf(ssta.lon,ssta.lat,linreg.p_value,
                 levels=[0, 0.05, 1],
                 zorder=1,
                 hatches=['..', None],
                 colors="None",
                 transform=ccrs.PlateCarree())
xtk = np.arange(-180,181,60)
ax.set_xticks(xtk)
# ax.set_xticks(xtk,crs=ccrs.PlateCarree())
ax.set_yticks(np.arange(-50,51,20),crs=ccrs.PlateCarree())
ax.yaxis.set_major_formatter(LatitudeFormatter())
ax.xaxis.set_major_formatter(LongitudeFormatter(zero_direction_label=True))
ax.xaxis.set_minor_locator(MultipleLocator(10))
ax.yaxis.set_minor_locator(MultipleLocator(5))
ax.coastlines()
ax.set_aspect("auto")
plt.colorbar(m)

```

**So troublesome!!!**

But if you `import sacpy.Map`, you can easily write:

```Python
import sacpy.Map
import cartopy.crs as ccrs
fig = plt.figure(figsize=[7, 3])
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
lon ,lat = ssta.lon , ssta.lat
m = ax.scontourf(lon,lat,rvalue)
n = ax.sig_plot(lon,lat,p,color="k",marker="..")
ax.init_map(stepx=50, ysmall=2.5)
plt.colorbar(m)
```

How wonderful, how concise !

## Acknowledgements

Thank Prof. Feng Zhu (NUIST,https://fzhu2e.github.io/) for his guidance of this project.

Thank for Prof. Tim Li (University of Hawaii at Mānoa, http://iprc.soest.hawaii.edu/people/li.php) ,Prof. Lin Chen (NUIST, https://faculty.nuist.edu.cn/chenlin12/zh_CN/index.htm) and Dr. Ming Sun (NUIST) 's help.

Sepcial thanks: Lifei Lin (Sun Yat-sen University) 's `repr_html.py` to visualize class in jupyter!


CHANGELOG.md
# Change Log

## version 0.0.1

Nothing!

## version 0.0.5

First official edition

## version 0.0.6

Updated the speed test and changed small errors

## version 0.0.7

Add examples and change README.md

## version 0.0.9

Add mult_corr,partial_corr in LinReg.py and spec_moth_dat, spec_moth_yrmean in XrTools.

Change examples.

## version 0.0.10

Add EOF analysis

## version 0.0.11

Change Example Data

## version 0.0.12

Fix a bug in XrTools

Add Map.py for quick plot

## version 0.0.13

Change example!

## version 0.0.14

Add SVD(MCA) analysis and visualization of LinReg in Jupyter.

## version 0.0.15

Add Map.py for test and new wrapper progress for Map.py and Linreg

## version 0.0.16

Optimized support for xarray

## version 0.0.17

1. update mask function for LinReg 
2. add shp mask function
3. correct SVD

## version 0.0.18

1. add new function of EOF
2. correct some errors of EOF and SVD 

## version 0.0.19 

1. add new function for Map
2. fix some errors of Map

sacpy/Map.py
try:
    from cartopy.mpl.geoaxes import GeoAxesSubplot
    import cartopy.crs as ccrs
    from cartopy.mpl.ticker import LongitudeFormatter, LatitudeFormatter
except:
    raise ImportError("Can't import cartopy, please check your envs")
from matplotlib.ticker import MultipleLocator
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import xarray as xr
from .Util import _correct_type
import matplotlib as mpl

mpl.rcParams['font.family'] = 'Times New Roman'
mpl.rcParams['image.cmap'] = 'RdBu_r'
# set font to TNR
# plt.rc('font', family='Times New Roman')
# plt.set_cmap("RdBu_r")


def get_levels(data, percentile: int = 98, num_level: int = 13, zero_sym: bool = True) -> np.ndarray:
    """ get levels of data

    Args:
        data (np.ndarray): data need to get levels
        percentile (int): max value or min value percentile
        num_level (int): colorbar levels
        zero_sym (bool): levels is zero symmetric or not

    Returns:
        np.ndarray: levels for contourf or contourf
    """
    #　set default values
    if percentile is None:
        percentile = 98
    if num_level is None:
        num_level = 13
    # if zero_sym is None:
    #     zero_sym = True
    # get min and max
    _max = np.nanpercentile(data, percentile)
    _min = np.nanpercentile(data, 100 - percentile)
    if zero_sym is None:
        if _min * _max < 0:
            zero_sym = True
        else:
            zero_sym = False
    # zero_sym :
    if zero_sym:
        max_real = np.max(np.abs([_max, _min]))
        levels = np.linspace(-max_real, max_real, num_level)
    else:
        levels = np.linspace(_min, _max, num_level)
    return levels


def _store_range(self, x, y):
    """ store data range for x and y of data

    Args:
        x (np.ndarray): x 
        y (np.ndarray): y
    """
    x = _correct_type(x)
    y = _correct_type(y)
    if x is not None:
        # save to xrange
        if not hasattr(self, "xrange"):
            self.xrange = [x]
        else:
            self.xrange.append(x)
    if y is not None:
        # save to yrange
        if not hasattr(self, "yrange"):
            self.yrange = [y]
        else:
            self.xrange.append(y)


def _get_extend(self):
    """ get self.xrange's xmin ,xmax and yrange's ymin and ymax
    Returns:
        xmin, xmax, ymin, ymax (float)
    """
    if hasattr(self, "xrange") and hasattr(self, "yrange"):
        xmin0, ymin0 = 360., 90.
        xmax0, ymax0 = -180, -90
        for xra, yra in zip(self.xrange, self.yrange):
            xmin1 = xra.min()
            ymin1 = yra.min()
            xmax1 = xra.max()
            ymax1 = yra.max()
            if xmin1 <= xmin0:
                xmin0 = xmin1
            if xmax1 >= xmax0:
                xmax0 = xmax1
            if ymin0 >= ymin1:
                ymin0 = ymin1
            if ymax0 <= ymax1:
                ymax0 = ymax1
        xmin, ymin, xmax, ymax = xmin0, ymin0, xmax0, ymax0
    else:
        xmin, ymin, xmax, ymax = None, None, None, None
    return xmin, xmax, ymin, ymax


def _get_extra_param(kwargs):
    """ get extra parameter for get levels

    Args:
        kwargs (dict)

    Returns:
        percentile, num_level, zero_sym (int,float,bool)
    """
    percentile = kwargs.get("percentile")
    if percentile is not None:
        del kwargs['percentile']
    num_level = kwargs.get("num_level")
    if num_level is not None:
        del kwargs['num_level']
    zero_sym = kwargs.get("zero_sym")
    if zero_sym is not None:
        del kwargs['zero_sym']
    return percentile, num_level, zero_sym


def _anal_args(args, clas="contourf"):
    """ analysis args for draw function

    Args:
        args (list)
        clas (str, optional): class of the function. Defaults to "contourf".

    Returns:
        np.ndarray: need value
    """
    if clas == "contourf":
        if len(args) == 1:
            levels0 = None
            Z = args[0]
            x, y = np.arange(Z.shape[0]), np.arange(Z.shape[1])
        elif len(args) == 2:
            levels0 = args[1]
            Z = args[0]
            x, y = np.arange(Z.shape[0]), np.arange(Z.shape[1])
        elif len(args) == 3:
            levels0 = None
            Z = args[2]
            x, y = args[0], args[1]
        elif len(args) == 4:
            levels0 = args[3]
            Z = args[2]
            x, y = args[0], args[1]
        else:
            raise ValueError("More parameters are transported")
        x = _correct_type(x)
        y = _correct_type(y)
        Z = _correct_type(Z)
        return Z, levels0, x, y
    elif clas == "pcolormesh":
        if len(args) == 1:
            Z = args[0]
            x, y = np.arange(Z.shape[0]), np.arange(Z.shape[1])
        elif len(args) == 2:
            raise ValueError("More parameters (2) are transported")
        elif len(args) == 3:
            Z = args[2]
            x, y = args[0], args[1]
        else:
            raise ValueError("More parameters are transported")
        x = _correct_type(x)
        y = _correct_type(y)
        Z = _correct_type(Z)
        return Z, x, y
    elif clas == "quiver":
        if len(args) == 1:
            raise ValueError("Need two arrays to depict wind")
        elif len(args) == 2:
            U, V = args[0], args[1]
            x, y = np.arange(U.shape[0]), np.arange(U.shape[1])
            C = None
        elif len(args) == 3:
            U, V = args[0], args[1]
            C = args[2]
            x, y = np.arange(U.shape[0]), np.arange(U.shape[1])
        elif len(args) == 4:
            x, y, U, V = args
            C = None
        elif len(args) == 5:
            x, y, U, V, C = args
        x = _correct_type(x)
        y = _correct_type(y)
        U = _correct_type(U)
        V = _correct_type(V)
        return x, y, U, V, C


def _contourf(self, *args, **kwargs):
    """ new contourf for GeoAxesSubplot

    Returns:
        mpl.contour.QuadContourSet
    """
    levels1 = kwargs.get("levels")
    extend = kwargs.get("extend")
    if extend is None:
        extend = "both"
        kwargs["extend"] = extend
    percentile, num_level, zero_sym = _get_extra_param(kwargs)
    Z, levels0, x, y = _anal_args(args)
    self._store_range(x, y)
    if levels0 is None and levels1 is None:
        levels = get_levels(Z, percentile, num_level, zero_sym)
        kwargs['levels'] = levels
    cmap = kwargs.get("cmap")
    if cmap is None:
        cmap = "RdBu_r"
        kwargs["cmap"] = cmap
    transform = kwargs.get("transform")
    if transform is None:
        transform = ccrs.PlateCarree()
        kwargs["transform"] = transform
    m = self.contourf(*args, **kwargs)
    return m


def _draw_ticks(self, extend, stepx=None, stepy=None, smallx=None, smally=None, bigx=None, bigy=None):
    """ draw map ticks

    Args:
        extend (list): extend of map [x1, x2, y1, y2]
        stepx (int, optional): big x step. Defaults to 30.
        stepy (int, optional): big y step. Defaults to 10.
        smallx (int, optional): x small step. Defaults to 5.
        smally (int, optional): y small step. Defaults to 5.
        bigx/bigy (int, optional): Resolution in X and Y directions
    """
    intend_ls = np.array([2, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60])
    [x1, x2, y1, y2] = extend
    delta_x = (x2 - x1) / intend_ls
    delta_y = (y2 - y1) / intend_ls
    if stepx is None:
        idx = np.argmin(np.abs(delta_x - 7))
        stepx = intend_ls[idx]
    if stepy is None:
        idy = np.argmin(np.abs(delta_y - 7))
        stepy = intend_ls[idy]
    if smallx is None:
        smallx = stepx / 5 if stepx != 4 else stepx / 5
    if smally is None:
        smally = stepy / 5 if stepy != 4 else stepy / 5
    if bigx is None:
        bigx = 10 if stepx != 2 else 5
    if bigy is None:
        bigy = 10 if stepy != 2 else 5
    # if stepx is None:
    #     stepx

    # autoly calculate stepx,stepy,smallx,smally,bigx,bigy
    # cling to 10 times
    xs = x1 // bigx * bigx if x1 % bigx == 0 else (x1 // bigx + 1) * bigx
    xe = x2 // bigx * bigx
    ys = y1 // bigy * bigy if y1 % bigy == 0 else (y1 // bigy + 1) * bigy
    ye = y2 // bigy * bigy
    # get xticks
    xticks = np.arange(xs, xe + 1, stepx)
    yticks = np.arange(ys, ye + 1, stepy)
    # set ticks
    self.set_xticks(xticks, crs=ccrs.PlateCarree())
    self.set_yticks(yticks, crs=ccrs.PlateCarree())
    self.yaxis.set_major_formatter(LatitudeFormatter())
    self.xaxis.set_major_formatter(LongitudeFormatter(zero_direction_label=True))
    self.xaxis.set_minor_locator(MultipleLocator(smallx))
    self.yaxis.set_minor_locator(MultipleLocator(smally))


def _initialize_map(self, same_size=True, coastlines=True, **kwargs):
    """ initial a map
    """
    # draw coastines
    if coastlines:
        self.coastlines()
    # Flatten map
    if same_size:
        self.set_aspect("auto")
    extend = self._get_extends()
    self.draw_ticks(extend, **kwargs)


def _contour(self, *args, **kwargs):
    """ new contour for GeoAxesSubplot

    Returns:
        mpl.contour.QuadContourSet
    """
    # get mpl.contour.QuadContourSet or not
    if not isinstance(args[0], (mpl.contour.QuadContourSet)):
        Z, levels0, x, y = _anal_args(args)
        self._store_range(x, y)
    # add transform
    transform = kwargs.get("transform")
    if transform is None:
        transform = ccrs.PlateCarree()
        kwargs['transform'] = transform
    n = self.contour(*args, **kwargs)
    return n


def _sig_ctrf(self, x, y, pvalue, thrshd=0.05, marker="..", color=None):
    """ Significance test dot

    Args:
        x (np.ndarray): x
        y (np.ndarray): y
        pvalue (np.ndarray): p value
        thrshd (float, optional): threshold of pvalue. Defaults to 0.05.
        marker (str, optional): mark of Significance test dot. Defaults to "..".
    """
    res = self.contourf(x,
                        y,
                        pvalue,
                        levels=[0, thrshd, 1],
                        zorder=1,
                        hatches=[marker, None],
                        colors="None",
                        transform=ccrs.PlateCarree())
    # set color
    if color is not None:
        for collection in res.collections:
            collection.set_linewidth(0.)
            collection.set_edgecolor(color)
    return res


def _sig_ctrf1(self, x, y, pvalue, thrshd=0.05, marker="..", color=None):
    """ Significance test dot

    Args:
        x (np.ndarray): x
        y (np.ndarray): y
        pvalue (np.ndarray): p value
        thrshd (float, optional): threshold of pvalue. Defaults to 0.05.
        marker (str, optional): mark of Significance test dot. Defaults to "..".
    """
    res = self.contourf(x,
                        y,
                        pvalue,
                        levels=[0, thrshd, 1],
                        zorder=1,
                        hatches=[marker, None],
                        colors="None")
    # set color
    if color is not None:
        for collection in res.collections:
            collection.set_linewidth(0.)
            collection.set_edgecolor(color)
    return res

def _pcolormesh(self, *args, **kwargs):
    """ new pcolormesh
    """
    Z, x, y = _anal_args(args, clas="pcolormesh")
    # store x ,y for ticks
    self._store_range(x, y)
    percentile, num_level, zero_sym = _get_extra_param(kwargs)
    vmin, vmax = kwargs.get("vmin"), kwargs.get("vmax")
    if vmin is None and vmax is None:
        levels = get_levels(Z, percentile, num_level, zero_sym)
        kwargs['vmin'] = levels[0]
        kwargs['vmax'] = levels[-1]
    cmap = kwargs.get("cmap")
    if cmap is None:
        cmap = "RdBu_r"
        kwargs["cmap"] = cmap
    transform = kwargs.get("transform")
    if transform is None:
        transform = ccrs.PlateCarree()
        kwargs["transform"] = transform
    m = self.pcolormesh(*args, **kwargs)
    return m


def _quiver(self, *args, **kwargs):
    """ new quiver
    """
    # add projcetion
    transform = kwargs.get("transform")
    if transform is None:
        transform = ccrs.PlateCarree()
        kwargs["transform"] = transform
    stepx = kwargs.get("stepx")
    if stepx is None:
        stepx = 1
    else:
        del kwargs['stepx']
    stepy = kwargs.get("stepy")
    if stepy is None:
        stepy = 1
    else:
        del kwargs['stepy']
    x, y, U, V, C = _anal_args(args, clas="quiver")
    self._store_range(x, y)
    if C is None:
        args1 = [x[::stepx], y[::stepy], U[::stepy, ::stepx], V[::stepy, ::stepx]]
    else:
        args1 = [x[::stepx], y[::stepy], U[::stepy, ::stepx], V[::stepy, ::stepx], C]
    m = self.quiver(*args1, **kwargs)
    return m


def _xr_splot(self, ax=None, label=0, kw1={}, kw2={}):
    if label == 0:
        lon = "lon"
        lat = "lat"
    else:
        lon = "longitude"
        lat = "latitude"
    if ax is None:
        ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=self[lon].mean().item()))
    m = ax.scontourf(self[lon], self[lat], self.to_numpy(), **kw1)
    ax.init_map(**kw2)
    plt.colorbar(m)
    return m, ax


GeoAxesSubplot.scontourf = _contourf
GeoAxesSubplot.scontour = _contour
GeoAxesSubplot.draw_ticks = _draw_ticks
GeoAxesSubplot.init_map = _initialize_map
GeoAxesSubplot.sig_plot = _sig_ctrf
GeoAxesSubplot.spcolormesh = _pcolormesh
GeoAxesSubplot._store_range = _store_range
GeoAxesSubplot._get_extends = _get_extend
GeoAxesSubplot.squiver = _quiver
xr.DataArray.splot = _xr_splot
mpl.axes.Axes.sig_plot = _sig_ctrf1
